### PR TITLE
Make sequence views compatible with var<matrix>

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This adds `.val()` methods to `scalar_seq_view<>` and `vector_seq_view<>` so that the probability distributions can work with `var<Matrix> types. Adding a `.val()` method requires access for `value_of()` so they both had to be moved to `fun` which requires a lot of copy/paste for their headers in the prob files. 

The changes in the prob functions are from just doing a regex replacing `value_of((.*)[(.*)])` with `$1.val($2)`

## Tests

tests added for the new `.val()` method in the rev testing folder.

## Side Effects

Nope

## Release notes

Adds sequence views for `var<matrix>` to make some of the distributions var matrix compatible

## Checklist

- [x] Math issue #1805 

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
@SteveBronder yoyo is this one suitable for me to review? Aka no fancy forwarding and whatnot? I want the sequence views for another pull.

File: test/unit/math/rev/meta/scalar_seq_view_test.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#include <stan/math/rev.hpp>
```
Comment (bbbales2) [inline test/unit/math/rev/meta/scalar_seq_view_test.cpp:1]: Missing tests for size and data and `operator[]`.

File: test/unit/math/rev/meta/vector_seq_view_test.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#include <stan/math/rev.hpp>
+#include <gtest/gtest.h>
+#include <vector>
+
+template <typename T>
+void test_vec_seq_var(const T& m1) {
```
Comment (bbbales2) [inline test/unit/math/rev/meta/vector_seq_view_test.cpp:6]: Missing tests for `size` and `operator[]`

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Review!

**Comment by SteveBronder** [comment]
Sorry need a lil more time before this is ready for review. On my 2 core little laptop so I have to use the jenkins to run the distribution tests etc.

**Comment by bbbales2** [comment]
Lolol, yeah, that's rough. Same situation here.

File: stan/math/prim/fun/vector_seq_view.hpp [hunk 3]
```diff
@@ -33,17 +33,33 @@ class vector_seq_view {};
...
+  static constexpr auto size() { return 1; }
+  inline const auto& operator[](size_t /* i */) const noexcept { return m_; }
+
+  template <typename C = T, require_st_arithmetic<C>* = nullptr>
+  inline const auto& val(size_t /* i */) const noexcept {
+    return m_;
+  }
+
+  template <typename C = T, require_st_autodiff<C>* = nullptr>
+  inline auto val(size_t /* i */) const noexcept {
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/fun/vector_seq_view.hpp:48]: For `val()` for a matrix type we'll return back an expression of the value's in the `Matrix<var>` or the values in `var<Matrix>` for `std::vector` this def needs to be a ref tho

File: stan/math/prim/prob/beta_binomial_lpmf.hpp [hunk 4]
```diff
@@ -159,7 +159,7 @@ return_type_t<T_size1, T_size2> beta_binomial_lpmf(const T_n& n, const T_N& N,
           += digamma_n_plus_alpha[i] + digamma_diff[i] - digamma_alpha[i];
     if (!is_constant_all<T_size2>::value)
       ops_partials.edge2_.partials_[i]
-          += digamma(value_of(N_vec[i] - n_vec[i] + beta_vec[i]))
+          += digamma(N_vec.val(i) - n_vec.val(i) + beta_vec.val(i))
```
Comment (SteveBronder) [inline stan/math/prim/prob/beta_binomial_lpmf.hpp:162]: Note this is one place where I did fix something by hand. I this supposed to be taking these without val wrapped around them or was that a typo?

File: stan/math/prim/fun/vector_seq_view.hpp [hunk 5]
```diff
@@ -55,11 +71,23 @@ class vector_seq_view<T, require_eigen_t<T>> {
...
+
+  inline const auto& operator[](size_t i) const { return v_[i]; }
+
+  template <typename C = T, require_st_arithmetic<C>* = nullptr>
+  inline const auto& val(size_t i) const {
+    return v_[i];
+  }
+
+  template <typename C = T, require_st_autodiff<C>* = nullptr>
+  inline auto val(size_t i) const {
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/fun/vector_seq_view.hpp:88]: `value_of()` here will return a new matrix in the case of an std vector and an expression of the values for a matrix

File: stan/math/prim/prob/beta_binomial_lpmf.hpp (hunk 4 above)
Comment (bbbales2) [reply stan/math/prim/prob/beta_binomial_lpmf.hpp:162]: This is a good fix. Doing the math before the value_of will create unnecessary varis if beta is autodiff.

File: stan/math/prim/fun/vector_seq_view.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/fun/vector_seq_view.hpp:48]: Oooh, this is the values of an autodiff type. Return by value makes sense here.

File: stan/math/prim/fun/scalar_seq_view.hpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+   */
+  inline auto operator[](size_t i) const { return c_.val().coeffRef(i); }
+  inline const auto* data() const noexcept { return c_.vi_; }
+  inline auto* data() noexcept { return c_.vi_; }
+
+  inline auto size() const noexcept { return c_.size(); }
+
+  template <typename T = C, require_st_arithmetic<T>* = nullptr>
+  inline auto val(size_t i) const {
+    return c_.val().coeffRef(i);
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/fun/scalar_seq_view.hpp:73]: Ooo `T` is not the `value_type` of the `var_value` ooof. Yeah this makes sense.

File: test/unit/math/rev/meta/vector_seq_view_test.cpp (hunk 2 above)
Comment (bbbales2) [reply test/unit/math/rev/meta/vector_seq_view_test.cpp]: I added a couple size checks [here](https://github.com/stan-dev/math/pull/2292/commits/d7124dff41dfeb7bc5fb58816d4658b21c8fdd78).

**Review by bbbales2 — APPROVED** [review]

**Comment by bbbales2** [comment]
This looks good to me. Tests were green this morning. Is this ready to merge? I added a couple size checks to vector sequence view and merged develop.

**Comment by SteveBronder** [comment]
Let me give this one more look over to make sure I didn't do anything stupid in the distributions

**Pull request merged**